"Logging on command" feature added

dseptem opened the pull request:
As promised, added the complete "logging on command" feature.
Can be set to True or False in the [chanlogs2] section of the sopel config.
Is off by default, logging all channels.
When turned on, the bot will start logging once it receives a command '.logging start' from its owner, and will stop once it receives '.logging stop'. Only accepts this command via channel messages.

dseptem committed:
on_command feature added

code is elegant now, on_command is off by default but can be turned on
to set enable logging on command in a channel per channel basis
on_command requires to be used on a channel and only acknowledged the
bot's owner
on_command usage example included for ".help logging"

dseptem committed:
minor typo fix

maxpowa commented on sopel_modules/chanlogs2/chanlogs2.py:
```diff
@@ -149,6 +153,21 @@ def redirect_topic(bot, trigger):
 
 
+@require_chanmsg('.logging is only permitted in channels')
+@require_owner("Sorry, I can't do that for you")
```
Since this is usually something that channel owners would want to have control over, maybe `require_op` would be better.

dseptem replied:
But what if the bot's owner is not OP?
This would hand over control of your own bot to other people.

What about @require_admin? This way, the owner could still command the bot while sharing it's control with the desired channel OPs.

maxpowa commented on sopel_modules/chanlogs2/chanlogs2.py:
```diff
@@ -149,6 +153,21 @@ def redirect_topic(bot, trigger):
 
+@require_chanmsg('.logging is only permitted in channels')
+@require_owner("Sorry, I can't do that for you")
+@module.commands("logging")
```
Should probably be consistent here, either everything `module` prefixed or nothing.

dseptem replied:
I didn't notice, sorry. Changing it to `module` prefixed,

maxpowa commented on sopel_modules/chanlogs2/chanlogs2.py:
```diff
@@ -149,6 +153,21 @@ def redirect_topic(bot, trigger):
+@require_chanmsg('.logging is only permitted in channels')
+@require_owner("Sorry, I can't do that for you")
+@module.commands("logging")
+@example('.logging (start|stop)')
```
Examples are meant to be directly runnable rather than showing syntax like this, they can even be used for running automated tests on the command (but that's a topic for another day).

dseptem replied:
Is it possible to add two examples? I tried it, but on .help call, it always showed the same one, leaving one forever hidden to the user.

maxpowa commented on sopel_modules/chanlogs2/chanlogs2.py:
```diff
@@ -162,16 +181,32 @@ def process_event(bot, trigger):
+        log_channel = bot.db.get_channel_value(channel, 'log_channel')
+    except:
+        log_channel = False
+    if bot.config.chanlogs2.on_command and log_channel:  # Checks if the bot is set to log on command and if the channel has to be logged
```
I prefer to `if condition: return` instead of wrapping the method body in an if statement, makes the flow a bit more clear. In this case it also has the added bonus of removing code duplication.

```
if config says per-channel toggle is on
    get logging value for channel from db
    return if value is nonexistent/off

do check nick/instance/backend
```

dseptem replied:
I think I got your idea, gonna implement it.

maxpowa commented on sopel_modules/chanlogs2/chanlogs2.py:
```diff
@@ -28,6 +30,7 @@ class Chanlogs2Section(StaticSection):
     # The following config options are only valid if the backend is 'file'
     logdir = FilenameAttribute('logdir', directory=True, default='~/chanlogs')
     by_day = ValidatedAttribute('by_day', parse=bool, default=True)
+    on_command = ValidatedAttribute('on_command', parse=bool, default=False)
```
`on_command` may be a bit confusing as it's not strictly logging the area around the command itself, perhaps `allow_toggle` or `op_toggle` would be a bit better.

dseptem replied:
Ok! Gonna change it to allow_toggle.

maxpowa commented on sopel_modules/chanlogs2/chanlogs2.py:
```diff
@@ -149,6 +153,21 @@ def redirect_topic(bot, trigger):
+@example('.logging (start|stop)')
+def logging_command(bot, trigger):
+    if trigger.group(2) == 'start':
+        bot.db.set_channel_value(trigger.sender, 'log_channel', True)
```
`log_channel` might not be the best name for the key here, perhaps just `logging` would be better, that way it appears (symbolically) as `#chan | logging | True` in the database.

dseptem replied:
But it would shadow the very own command of this feature, `logging`.
Maybe the command should be changed to `.log start` and `.log stop` ?

maxpowa reviewed (commented):
Mainly opinionated stuff, but some things are stylistic.

dseptem committed:
Fixes according to suggestions

-renamed from on_command to allow_toggle
-changed from require_owner to require_admin
-consistent usage of module prefix
-changed example to ".log start"
-renamed from '.logging start/stop' to '.log start/stop'
-remade write_log if condition according to suggestion
